### DISCUSSION TITLE
React com TDD

### DISCUSSION THREAD

**Discussion by phsantiago** [frontendbr/forum#729] [Perguntas e Respostas]
Meu primeiro dia com TDD real mesmo, feito da maneira certa e oh shit não tá fácil isso. Parece que não sei pensar como as coisas devem ser feita.
Alguém já teve experiência usando TDD com React? 

Quero criar alguns testes de comportamento antes de desenvolver um App de calendário. Está sendo muito mind blowing pensar que tenho que escrever o teste antes mesmo de desenvolver os componentes, por exemplo, se eu quiser testar se o calendário troca o mês corretamente, tenho que testar se o usuário consegue chegar até ele pela rota primeiro?

Alguém tem experiência nesse tópico e pode dar algumas dicas, sobre TDD mesmo ou sobre TDD com React?

**Comment by ghost** [comment]
Cara um bom ponto de partida é o curso de [TDD com JS](https://www.udemy.com/js-com-tdd-na-pratica) do @willianjusten ou o curso do @fdaciuk de [React](https://www.udemy.com/curso-reactjs-ninja) que cobre Jest e TDD, pra começar tente entender o conceito de teste unitário e as diversas metodologias, comece por TDD e você vai ver que não é um bicho de 7 cabeças. Testar componentes React também depende de quanto de mutação de estado seu componente faz, se você construir componentes puros a tendência é que a complexidade diminua.

**Comment by doug2k1** [comment]
Tem cenários que o TDD real, _by the book_, não é tão simples de fazer. Nesses casos eu procuro pelo menos pensar nos testes enquanto desenvolvo, assim o código naturalmente já mais mais testável. 

Pra fazer **Desenvolvimento Orientado a Testes** os testes não precisam necessariamente estar escritos, podem estar na sua cabeça. Só procure ir intercalando, desenvolver um pouco, e escrever um pouco de teste, e quando der, escrever o teste antes. Não deixe acumular muito, senão fica mais difícil.

**Comment by fdaciuk** [comment]
@phsantiago uma coisa importante que você deve ter em mente: testes são _uma forma de pensar_. Se você nunca escreveu um teste automatizado antes, e está fazendo isso direto em um projeto, dificilmente você vai conseguir seguir sozinho, pois as dúvidas são **normais**, pois existem vários tipos de testes que você pode fazer: unitário, de integração, de aceitação, E2E (end-to-end ou testes de tela), etc.

O ponto principal não é nem conhecer todos os tipos de testes, mas inicialmente você precisa começar do começo =)

Recomendo que você comece estudar sobre testes unitários, e comece a fazer TDD com funções puras, pois é o que você normalmente vai fazer no seu dia a dia quando estiver usando TDD de verdade.

Procure grupos de estudo perto de onde você mora, meetups, etc., onde você tenha contato com outras pessoas que também estudam sobre testes. Esse contato vai facilitar o seu aprendizado, e vai ficar mais fácil você entender o porquê de aplicar testes automatizados.

Por agora posso te recomendar alguns artigos que irão te ajudar e dar uma introdução a esse mundo de testes:

Artigo: https://blog.da2k.com.br/2015/01/06/pensando-tdd-com-javascript/
Artigo: https://blog.da2k.com.br/2015/01/07/code-coverage-em-javascript/
Série (leia todos os artigos): https://blog.da2k.com.br/2015/03/05/javascript-entendendo-e-criando-suas-proprias-promises/

Artigo: https://willianjusten.com.br/entendendo-testes-de-software/

Comece seguindo esses artigos, e vá tentando escrever testes aos poucos. Tente entender essa mentalidade de testes, que vai ficar muito mais fácil de você decidir quando e qual tipo de teste aplicar ao tipo de aplicação que você está desenvolvendo :)

**Comment by phsantiago** [comment]
@fdaciuk Ótima contribuição, mudou minha maneira de pensar. Estou lendo todos os artigos, obrigado.

**Comment by willianjusten** [comment]
Opa, adicionando mais um link, há pouco eu adicionei um post criando um componente React com TDD do zero e bem na prática mesmo de Teste => Código => Refactor.

Talvez você se interesse a ler também:

https://willianjusten.com.br/criando-componentes-react-com-testes-lyef-flag/

**Comment by EdsonECS** [comment]
Não tem jeito, no começo é sempre assim um componente React com TDD, como o Willianjusten comentou.
Muito interessante.

https://blog.da2k.com.br/2015/01/07/code-coverage-em-javascript/

**Comment by willianjusten** [comment]
Um tempinho atrás eu também fiz uns [videos bem atuais criando uma lib com testes do zero](https://willianjusten.com.br/criando-uma-lib-js-com-testes-do-zero/)

E também tem um outro vídeo que [refatoro uma lib com testes e react hooks](https://willianjusten.com.br/convertendo-um-class-based-component-para-react-hooks/)

Espero que seja útil.

**Comment by phsantiago** [comment]
Haha que loucura, quase 2 anos que criei essa issue e hoje eu trabalho com a primeira pessoa que respondeu ela, o @edmolima! 

@willianjusten parabéns pelo material, e pelo mindset de refatorar com testes, vejo poucas pessoas no mercado fazendo isso hoje!

**Comment by phsantiago** [comment]
Vai rolar um webinar do Eric Eliot sobre TDD, ele vai falar sobre componentes também https://tddday.com/ só que ta 120 doletas. Mesmo sendo um conteúdo que eu já tenho um bom domínio to pensando em pagar.

> “It is impossible for a man to learn what he thinks he already knows.”
― Epictetus

**Comment by felipefialho** [comment]
@phsantiago Sensacional haha 😄
